packager/certloader/certloader.go

```diff
@@ -17,33 +17,89 @@ package certloader
 import (
 	"crypto"
 	"crypto/x509"
+	"encoding/pem"
 	"io/ioutil"
 	"log"
+	"os"
 
 	"github.com/WICG/webpackage/go/signedexchange"
+	"github.com/gofrs/flock"
 	"github.com/pkg/errors"
 
-	"github.com/ampproject/amppackager/packager/certcache"
+	"github.com/ampproject/amppackager/packager/certfetcher"
 	"github.com/ampproject/amppackager/packager/util"
 )
 
-// Creates cert cache by loading certs and keys from disk, doing validation
-// and populating the cert cache with current set of certificate related information.
-// If development mode is true, prints a warning for certs that can't sign HTTP exchanges.
-func PopulateCertCache(config *util.Config, key crypto.PrivateKey, developmentMode bool) (*certcache.CertCache, error) {
-	certs, err := loadCertsFromFile(config, developmentMode)
-	if err != nil {
-		return nil, err
+func CreateCertFetcher(config *util.Config, key crypto.PrivateKey, domain string,
+	developmentMode bool, autoRenewCert bool) (*certfetcher.CertFetcher, error) {
+	if !autoRenewCert {
+		// Certfetcher can be nil, if auto renew is off.
+		return nil, nil
+	}
+
+	if config.ACMEConfig == nil {
+		return nil, errors.New("missing ACMEConfig")
+	}
+	if config.ACMEConfig.Production == nil {
+		return nil, errors.New("missing ACMEConfig.Production")
 	}
-	for _, urlSet := range config.URLSet {
-		domain := urlSet.Sign.Domain
-		if err := util.CertificateMatches(certs[0], key, domain); err != nil {
-			return nil, errors.Wrapf(err, "checking %s", config.CertFile)
+	if config.ACMEConfig.Production.EmailAddress == "" {
```

Comment:
Typically I've put this kind of validation in util/config.go. (Pro is the definition of what's a valid config is in one place; con is it's farther away from the thing actually using it.) WDYT?